### PR TITLE
`strict-match: ON` in a host entry makes it ineligible for use as the fallback host entry.

### PR DESCRIPTION
**Pull request by jhatala**
Requests without :authority (or a Host header field) and requests that do not match any of the configured host entries will be routed to the first host entry that does not have the strict-match flag.  If all entries have the flag, then a request that does not match any of them will get a 404 "not found".

This allows to list specific entries first before a wildcard entry and still to route requests without Host (or :authority) to the wildcard entry.

Example:

```
hosts:
  "www.example.com":
    strict-match: ON
    paths:
      /:
        file.dir: examples/doc_root1
  "*.example.com":
    strict-match: ON
    paths:
      /:
        file.dir: examples/doc_root2
  "*":
    paths:
      /:
        file.dir: examples/doc_root3
```

In this example, requests without a Host header will be routed to the third host entry.

Without a way to mark host entries as ineligible to be the default virtual host, requests without a Host header can only be routed to the first host entry in the configuration, and that is not always the appropriate destination for those requests.

### CODE REVIEW

**Review by kazuho — COMMENTED** [review]
Thank you for the changes. Looks great, and I really appreciate to see __increased__ coverage with the tests being added. Thank you very much.

I see that there's additional O(N) iteration when a matching is host not found. However, that is not of an immediate concern as typically the default host would be the first entry :-) If it becomes an issue, we can consider how to optimize this.

My comments below are just editorial nitpicks. WDYT?

File: include/h2o.h [hunk 1]
```diff
@@ -327,6 +335,11 @@ struct st_h2o_globalconf_t {
      * a NULL-terminated list of host contexts (h2o_hostconf_t)
      */
     h2o_hostconf_t **hosts;
+    /**
+     * The hostconf that will be used when none of the hostconfs for the
+     * listener match the request and they all have strict-match: ON.
+     */
+    h2o_hostconf_t *default_host;
```
[the remark this answers is not in the document]
Comment (jhatala) [reply include/h2o.h]: That makes sense, thank you.  Done in 83aa5a36b.

File: lib/core/configurator.c [hunk 2]
```diff
@@ -961,6 +972,9 @@ void h2o_configurator__init_core(h2o_globalconf_t *conf)
         h2o_configurator_define_command(
             c, "paths", H2O_CONFIGURATOR_FLAG_HOST | H2O_CONFIGURATOR_FLAG_EXPECT_MAPPING | H2O_CONFIGURATOR_FLAG_DEFERRED,
             on_config_paths);
+        h2o_configurator_define_command(
+            c, "strict-match", H2O_CONFIGURATOR_FLAG_HOST | H2O_CONFIGURATOR_FLAG_EXPECT_SCALAR | H2O_CONFIGURATOR_FLAG_DEFERRED,
```
[the remark this answers is not in the document]
Comment (jhatala) [reply lib/core/configurator.c]: Thank you.  Done in 83aa5a36b.

**Comment by kazuho** [comment]
Thank you for the changes. Merged to master.

**Pull request merged**